rpcs3/Emu/RSX/color_utils.h

```diff
@@ -0,0 +1,245 @@
+		blue = renormalize_color8(raw_value, 0x1f);
+		green = renormalize_color8(raw_value >> 5, 0x3f);
+		red = renormalize_color8(raw_value >> 11, 0x1f);
+	}
+
+	static inline void get_a1rgb555_clear_color(u8& red, u8& green, u8& blue, u8& alpha, u8 alpha_override)
+	{
+		// RSX clear color is just a memcpy, so in this case the input is ARGB8 so only BG have the 16-bit input
+		const u16 raw_value = static_cast<u16>(green) << 8 | blue;
+		blue = renormalize_color8(raw_value, 0x1f);
+		green = renormalize_color8(raw_value >> 5, 0x1f);
+		red = renormalize_color8(raw_value >> 10, 0x1f);
+
+		// Alpha can technically be encoded into the clear but the format normally just injects constants.
+		// Will require hardware tests when possible to determine which approach makes more sense.
+		// alpha = static_cast<u8>((raw_value & (1 << 15)) ? 255 : 0);
+		alpha = alpha_override;
+	}
+
+	static inline u32 get_b8_clearmask(u32 mask)
+	{
+		u32 result = 0;
+		if (mask & RSX_GCM_CLEAR_BLUE_BIT) result |= RSX_GCM_CLEAR_RED_BIT;
+		return result;
+	}
+
+	static inline void get_b8_colormask(bool& red, bool& green, bool& blue, bool& alpha)
+	{
+		red = blue;
+		green = false;
+		blue = false;
+		alpha = false;
+	}
+
+	static inline void get_b8_clear_color(u8& red, u8& /*green*/, u8& blue, u8& /*alpha*/)
+	{
+		std::swap(red, blue);
+	}
+
+	static inline color4f decode_border_color(u32 colorref)
+	{
+		color4f result;
+		result.b = (colorref & 0xFF) / 255.f;
+		result.g = ((colorref >> 8) & 0xFF) / 255.f;
+		result.r = ((colorref >> 16) & 0xFF) / 255.f;
+		result.a = ((colorref >> 24) & 0xFF) / 255.f;
+		return result;
+	}
+
+	static inline u32 encode_color_to_storage_key(color4f color)
+	{
+		const u32 r = static_cast<u8>(color.r * 255);
+		const u32 g = static_cast<u8>(color.g * 255);
+		const u32 b = static_cast<u8>(color.b * 255);
+		const u32 a = static_cast<u8>(color.a * 255);
+
+		return (a << 24) | (b << 16) | (g << 8) | r;
+	}
+
+	static inline const std::array<bool, 4> get_write_output_mask(rsx::surface_color_format format)
+	{
+		constexpr std::array<bool, 4> rgba = { true, true, true, true };
+		constexpr std::array<bool, 4> rgb = { true, true, true, false };
+		constexpr std::array<bool, 4> rg = { true, true, false, false };
+		constexpr std::array<bool, 4> r = { true, false, false, false };
+
+		switch (format)
+		{
+		case rsx::surface_color_format::a8r8g8b8:
+		case rsx::surface_color_format::a8b8g8r8:
+		case rsx::surface_color_format::w16z16y16x16:
+		case rsx::surface_color_format::w32z32y32x32:
+			return rgba;
+		case rsx::surface_color_format::x1r5g5b5_z1r5g5b5:
+		case rsx::surface_color_format::x1r5g5b5_o1r5g5b5:
+		case rsx::surface_color_format::r5g6b5:
+		case rsx::surface_color_format::x8r8g8b8_z8r8g8b8:
+		case rsx::surface_color_format::x8r8g8b8_o8r8g8b8:
+		case rsx::surface_color_format::x8b8g8r8_z8b8g8r8:
+		case rsx::surface_color_format::x8b8g8r8_o8b8g8r8:
+			return rgb;
+		case rsx::surface_color_format::g8b8:
+			return rg;

```

Comment:
why rg ?

Reply:
Because gb is not a valid texture format. They all start from R on PC